homeassistant/components/media_player/epson.py

```diff
@@ -0,0 +1,419 @@
+        hass.data[DATA_EPSON].append(epson)
+        async_add_devices([epson], update_before_add=True)
+
+    @asyncio.coroutine
+    def async_service_handler(service):
+        """Handle for services."""
+        entity_ids = service.data.get(ATTR_ENTITY_ID)
+        if entity_ids:
+            devices = [device for device in hass.data[DATA_EPSON]
+                       if device.entity_id in entity_ids]
+        else:
+            devices = hass.data[DATA_EPSON]
+        for device in devices:
+            if service.service == ATTR_CMODE:
+                cmode = service.data.get(ATTR_CMODE).lower()
+                if cmode in CMODE_LIST_SET:
+                    yield from device.select_cmode(cmode)
+            yield from device.update()
+    hass.services.async_register(
+        DOMAIN, ATTR_CMODE, async_service_handler,
+        schema=EPSON_SCHEMA)
+    return True
+
+
+class EpsonProjector(MediaPlayerDevice):
+    """Representation of Epson Projector Device."""
+
+    def __init__(self, hass, name, host, port, encryption, key_commands):
+        """Initialize entity to control Epson projector."""
+        self._hass = hass
+        self._name = name
+        self._host = host
+        self._port = port
+        self._cmode = None
+        self._source_list = list(DEFAULT_SOURCES.values())
+        self._source = None
+        self._volume = None
+        self.key_commands = key_commands
+        self._encryption = encryption
+        self._state = None
+        http_protocol = 'https' if self._encryption else 'http'
+        self._http_url = '{http_protocol}://{host}:{port}/cgi-bin/'.format(
+            http_protocol=http_protocol,
+            host=self._host,
+            port=self._port)
+        referer = "{http_protocol}://{host}:{port}/cgi-bin/webconf".format(
+            http_protocol=http_protocol,
+            host=self._host,
+            port=self._port)
+        self._headers = {
+            "Accept-Encoding": ACCEPT_ENCODING,
+            "Accept": CONTENT_TYPE_JSON,
+            "Referer": referer
+        }
+        self.websession = async_create_clientsession(
+            hass,
+            verify_ssl=False)
+
+    @asyncio.coroutine
+    def update(self):
+        """Update state of device."""
+        is_turned_on = yield from self.get_property('PWR')
+        if is_turned_on and is_turned_on == EPSON_CODES['ON']:
+            self._state = STATE_ON
+            cmode = yield from self.get_property('CMODE')
+            if cmode and cmode in CMODE_LIST:
+                self._cmode = CMODE_LIST[cmode]
+            source = yield from self.get_property('SOURCE')
+            if source and source in SOURCE_LIST:
+                self._source = SOURCE_LIST[source]
+            volume = yield from self.get_property('VOLUME')
+            if volume:
+                self._volume = volume
+        else:
+            self._state = STATE_OFF
+
+    @asyncio.coroutine
+    def get_property(self, command):
```

Comment:
This is too much protocol specific information. Please extract this into a third party library and have the platform use that. This should include all key mappings and cmodes etc